src/ParsingMiddleware.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Amp\Http\Server\FormParser;
+
+use Amp\Http\Server\ErrorHandler;
+use Amp\Http\Server\Middleware;
+use Amp\Http\Server\Request;
+use Amp\Http\Server\RequestHandler;
+use Amp\Http\Server\Server;
+use Amp\Http\Server\ServerObserver;
+use Amp\Http\Status;
+use Amp\Promise;
+use Amp\Success;
+use function Amp\call;
+
+class ParsingMiddleware implements Middleware, ServerObserver
+{
+    /** @var BufferingParser */
+    private $parser;
+
+    /** @var ErrorHandler */
+    private $errorHandler;
+
+    public function __construct(int $fieldCountLimit = null)
+    {
+        $this->parser = new BufferingParser($fieldCountLimit);
+    }
+
+    public function handleRequest(Request $request, RequestHandler $requestHandler): Promise
+    {
+        return call(function () use ($request, $requestHandler) {
+            try {
+                $request->setAttribute(self::class, yield $this->parser->parseForm($request));
```

Comment:
I think it should use `Form::class` as attribute key?

Reply:
Yeah, without the second middleware that makes more sense.